### PR TITLE
refactor: convert more files to typescript

### PR DESCRIPTION
**Pull request by MarshallOfSound**
I'm doing this work in small chunks to make review easier.  This is another chunk of TS conversion

Notes: no-notes

This PR also adds the `tsify` dependency to allow TS transformations for the `browserify`'ed code for sandbox and isolated

### CODE REVIEW

File: lib/browser/init.ts [hunk 1]
```diff
@@ -49,10 +49,12 @@ process.on('uncaughtException', function (error) {
   }
 
   // Show error in GUI.
-  const dialog = require('electron').dialog
-  const stack = error.stack ? error.stack : `${error.name}: ${error.message}`
-  const message = 'Uncaught Exception:\n' + stack
-  dialog.showErrorBox('A JavaScript error occurred in the main process', message)
+  import('electron')
```
Comment (nornagon) [inline lib/browser/init.ts]: why not a static import?

File: lib/browser/init.ts [hunk 2]
```diff
@@ -192,10 +194,13 @@ app.on('window-all-closed', () => {
   }
 })
 
-const { setDefaultApplicationMenu } = require('@electron/internal/browser/default-menu')
-
-// Create default menu.
-app.once('ready', setDefaultApplicationMenu)
+Promise.all([
+  import('@electron/internal/browser/default-menu'),
```
Comment (nornagon) [inline lib/browser/init.ts]: here too

File: lib/common/init.ts [hunk 3]
```diff
@@ -1,29 +1,35 @@
...
       return func.apply(this, arguments)
-    }
+    } as any
   })
 }
 
-function wrap (func, wrapper) {
+/**
+ * Casts to any below for func are due to Typescript not supporting symbols
+ * in index signitures
```
Comment (nornagon) [inline lib/common/init.ts]:
```suggestion
 * in index signatures
```

File: tsconfig.json [hunk 4]
```diff
@@ -1,23 +1,23 @@
...
+    "experimentalDecorators": false,
+    "strict": true,
+    "baseUrl": ".",
+    "allowJs": true,
+    "outDir": "ts-gen",
+    "paths": {
+      "@electron/internal/*": ["lib/*"]
+    }
+  },
+  "exclude": [
```
Comment (nornagon) [inline tsconfig.json]: can this be `include` instead of `exclude`?

File: lib/browser/init.ts (hunk 1 above)
Comment (MarshallOfSound) [reply lib/browser/init.ts]: You can't `import 'electron'` half way down a file, static imports have to be at the top of a file

File: lib/browser/init.ts (hunk 2 above)
Comment (MarshallOfSound) [reply lib/browser/init.ts]: Same as above, and the async nature here actually works well with `app.whenReady`

File: lib/browser/init.ts (hunk 1 above)
Comment (nornagon) [reply lib/browser/init.ts]: is there any particular reason not to just import this at the top of the file then?

File: tsconfig.json (hunk 4 above)
Comment (MarshallOfSound) [reply tsconfig.json]: Not easily, we have to explicitly `exclude` this file and this file only

Comment (MarshallOfSound) [reply tsconfig.json]: Otherwise TS get's multiple-type-declaration errors with the TS defs GN passes in

File: lib/browser/init.ts (hunk 1 above)
Comment (MarshallOfSound) [reply lib/browser/init.ts]: We can't 😢 The `electron` alias is configured / setup half-way down this file in `reset-search-paths`

Comment (nornagon) [reply lib/browser/init.ts]: ah, okay. Let's comment here about that then!

File: lib/common/init.ts [hunk 5]
```diff
@@ -1,29 +1,35 @@
...
 // called.
-const wrapWithActivateUvLoop = function (func) {
+const wrapWithActivateUvLoop = function <T extends Function> (func: T): T {
   return wrap(func, function (func) {
-    return function () {
+    return function (this: any) {
       process.activateUvLoop()
       return func.apply(this, arguments)
-    }
+    } as any
```
Comment (felixrieseberg) [inline lib/common/init.ts]: Can we do better than `any`? It seems like it'll at least be `(...args: Array<any>) => any`.

**Comment by CvX** [comment]
This PR (or more specifically https://github.com/electron/electron/pull/16820/files#diff-27aca9003e7bfe3e2c635e67d04fc917L17) should probably be back-ported to 5-0-x. See: https://github.com/electron/electron/issues/17647

**Pull request merged**